### ISSUE TITLE
Why is the behavior of "reparameterize" different in training and testing?

### ISSUE THREAD

**Issue by bcpenggh** [2017-fall-DL-training-program/VAE-GAN-and-VAE-GAN#22]
Dear TA,

In the VAE lab, the example code behave differently in training and testing. In training, a random number of specified mean and variance is returned. In testing, the mean is returned directly. Here are my questions.
1. Are the mean and variance the statistical values of the whole mini batch?
2. If (1) is true and all the photos in the batch are different, how can we control the decoder output to be similar to a specific photo? In other words, how can we minimize the loss of different photos?
3. In testing mode, the variance is ignored. If (1) is true, then all the generated photos are the same. So which part of my understanding is incorrect?

The code to reparameterize is shown below
    def reparameterize(self, mu, logvar):
        if self.training:
          std = logvar.mul(0.5).exp_()
          eps = Variable(std.data.new(std.size()).normal_())
          return eps.mul(std).add_(mu)
        else:
          return mu

**Comment by hui-po-wang** [comment]
Hi @bcpenggh,

1. The mean and variance are the statistical value of single data point. For example, given inputs with size (batch size, data_size), the encoder would generates a batch of mean with size (batch_size, latent code size) and a batch of variance with size (batch size, latent code size). Each of them is generated corresponding to a specific input data sample.

2. Same as above, each mean and variance are generated corresponding to specific single inputs.

3. The reason that we added variance during testing time is to release the constraints of latent codes and tell the decoder that please treat latent codes the same if they come from the same local area. Imagine that if it didn't have randomness (var), each latent code would be a discrete, sharp, and very specific point which may only represent one image. Thus, the decoder may have difficulty in reconstructing images even when the inputs of auto-encoder are just a little different from the seen inputs, because the latent codes are very sharp and it's very difficult to make two point exactly same in the continuous space.

Thanks

**Comment by bcpenggh** [comment]
Hi TA,

Let me try to use my words to repeat again. The encoder can generate the means and variances of a set of latent code of each photo. That is, the photo is not represented by a set of deterministic latent values, instead it is represented by a "region" specified by the means and variances. Around that region the decoder outputs the similar photo close to the original one.

In training mode, we reparameterize the set of latent codes of the specific photo so that we can make sure the decoder covers the whole "region." On the other hand in testing mode, we just need a set of latent code (a single point within that "region") to check the similarity of the generated photos. So we use the mean values and ignore the variances.

Do I understand correctly?

**Comment by hui-po-wang** [comment]
Hi @bcpenggh ,

Yes, good summary.

**Comment by bcpenggh** [comment]
Thanks a lot :)
